pgxn/neon/file_cache.c

```diff
@@ -801,16 +835,34 @@ lfc_readv_select(NRelFileInfo rinfo, ForkNumber forkNum, BlockNumber blkno,
 
 		if (iteration_hits != 0)
 		{
+			/* chunk offset (# of pages) into the LFC file */
+			off_t	first_read_offset = (off_t) entry_offset * BLOCKS_PER_CHUNK;
+			/* offset of first IOV */
+			first_read_offset += chunk_offs + first_block_in_chunk_read;
+
 			pgstat_report_wait_start(WAIT_EVENT_NEON_LFC_READ);
-			rc = preadv(lfc_desc, iov, blocks_in_chunk,
-						((off_t) entry_offset * BLOCKS_PER_CHUNK + chunk_offs) * BLCKSZ);
+
+			/* Read only the blocks we're interested in, limiting */
+			rc = preadv(lfc_desc, &iov[first_block_in_chunk_read],
+						iov_last_used - first_block_in_chunk_read,
+						first_read_offset * BLCKSZ);
 			pgstat_report_wait_end();
 
 			if (rc != (BLCKSZ * blocks_in_chunk))
```

Comment:
It is not correct because we are reading  `iov_last_used - first_block_in_chunk_read` rather than `blocks_in_chunk`